server/src/main/java/io/unitycatalog/server/service/Scim2UserService.java

```diff
@@ -0,0 +1,204 @@
+ *   <li>externalId - maps to SCIM external id
+ *   <li>userName - maps to SCIM primary email
+ * </ul>
+ */
+@ExceptionHandler(GlobalExceptionHandler.class)
+public class Scim2UserService {
+  private static final UserRepository USER_REPOSITORY = UserRepository.getInstance();
+
+  public Scim2UserService() {}
+
+  @Get("")
+  public HttpResponse getScimUsers(
+      @Param("filter") Optional<String> filter,
+      @Param("startIndex") Optional<Integer> startIndex,
+      @Param("count") Optional<Integer> count) {
+    final Filter userFilter = filter.filter(f -> !f.isEmpty()).map(this::parseFilter).orElse(null);
+    FilterEvaluator filterEvaluator = new FilterEvaluator();
+
+    List<UserResource> userResourcesList =
+        USER_REPOSITORY.listUsers().stream()
+            .map(this::asUserResource)
+            .filter(m -> match(filterEvaluator, userFilter, m))
+            .skip(startIndex.orElse(1) - 1)
+            .limit(count.orElse(50))
+            .toList();
+    return HttpResponse.ofJson(userResourcesList);
+  }
+
+  @Post("")
+  public HttpResponse createScimUser(UserResource userResource) {
+    // Get primary email address
+    Email primaryEmail =
+        userResource.getEmails().stream()
+            .filter(Email::getPrimary)
+            .findFirst()
+            .orElseThrow(
+                () ->
+                    new Scim2RuntimeException(
+                        new PreconditionFailedException("User does not have a primary email.")));
+
+    try {
+      User user = USER_REPOSITORY.getUserByEmail(primaryEmail.getValue());
+      return HttpResponse.ofJson(asUserResource(user));
+    } catch (BaseException e) {
+      if (e.getErrorCode() == ErrorCode.NOT_FOUND) {
+        String pictureUrl = "";
+        if (userResource.getPhotos() != null && !userResource.getPhotos().isEmpty()) {
+          pictureUrl = userResource.getPhotos().get(0).getValue().toString();
+        }
+        User user =
+            USER_REPOSITORY.createUser(
+                new CreateUser()
+                    .name(userResource.getDisplayName())
+                    .email(primaryEmail.getValue())
+                    .externalId(userResource.getExternalId())
+                    .pictureUrl(pictureUrl));
+        return HttpResponse.ofJson(asUserResource(user));
+      } else {
+        throw e;
+      }
+    }
+  }
+
+  @Get("/self")
+  public HttpResponse getCurrentUser() {
+    ServiceRequestContext ctx = ServiceRequestContext.current();
+    DecodedJWT decodedJWT = ctx.attr(AuthDecorator.DECODED_JWT_ATTR);
+    Claim sub = decodedJWT.getClaim("sub");
+    return HttpResponse.ofJson(asUserResource(USER_REPOSITORY.getUserByEmail(sub.asString())));
+  }
+
+  @Get("/{id}")
+  public HttpResponse getUser(@Param("id") String id) {
+    return HttpResponse.ofJson(asUserResource(USER_REPOSITORY.getUser(id)));
+  }
+
+  @Put("/{id}")
+  public HttpResponse updateUser(@Param("id") String id, UserResource userResource) {
```

Comment:
Nit: the update code does the read and modify in different DB transactions which can cause lost updates for concurrent updates.

Reply:
the repository UpdateUser was doing a bunch of this already so I simplified the code. Done.